### PR TITLE
UDP support + Disable ELB creation

### PR DESCRIPTION
**Pull request by spencerhakim**
First attempt at #665 + #492, I'd love to get some feedback as I go. I haven't tested any of it yet, but hopefully there won't be too many problems.

EDIT: OK, some notes. Since ELBs only support TCP, I treat ports as having one of 3 possible states: Public/External TCP, Internal TCP, or UDP. I also did my best to update any code that used something like `len(s.Ports) > 0` to determine if an ELB should be created to instead use `s.HasBalancer()`, and in general to add protocol checks to code that assumed ports would always be TCP. The regex I used for parsing the ports from the yaml file is forgiving and backwards compatible, while the regex used for parsing ports from the ECSTaskDefinition is not; if that was a mistake, I can easily correct it.

As a side-effect of adding UDP support, I was able to easily add in a check for `convox.elb=false` to disable ELB creation entirely, even when TCP ports are defined. Feel free to suggest a better name for this label.

I still need to add my own tests, but I've fixed the existing tests to match.

### CODE REVIEW

File: manifest/yaml.go [hunk 1]
```diff
@@ -271,41 +272,29 @@ func (pp *Ports) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
-				return fmt.Errorf("error parsing port: %s", err)
-			}
+		container, _ := strconv.Atoi(parts[2])
+		p.Container = container
+		p.Balancer = container
 
-			p.Name = parts[0]
-			p.Container = n
+		// Only TCP ports can be "public" (in the ELB sense) or have an ELB at all
+		if parts[1] != "" && p.Protocol == TCP {
```
Comment (spencerhakim) [inline manifest/yaml.go]: I debated with myself heavily on how to handle this. I think a proper `State` enum to encapsulate `Protocol` + `Public` and ensure that UDP ports are never accidentally set to public would be cleanest, but I opted for the less invasive option of simply adding `Protocol` and leaving `Public` untouched. I'm still debating whether or not to go back and refactor it.

File: manifest/service.go [hunk 2]
```diff
@@ -96,14 +96,24 @@ func (s *Service) Process(app string, m Manifest) Process {
 }
 
 func (s Service) HasBalancer() bool {
-	return len(s.Ports) > 0
+	if s.Labels["convox.elb"] == "false" {
```
Comment (ddollar) [inline manifest/service.go]: Would like to avoid AWS-specific naming here. Perhaps `convox.balancer`

File: manifest/yaml.go (hunk 1 above)
Comment (ddollar) [inline manifest/yaml.go]: Makes sense to me

Comment (spencerhakim) [inline manifest/yaml.go]: Sorry, just to clarify, you think it makes sense to go back and change it, or to leave it as is?

Comment (ddollar) [inline manifest/yaml.go]: I think this change makes sense as is.

**Comment by spencerhakim** [comment]
I added some new tests, and the diffs on the generated fixtures looked correct, though I'm hardly an expert on CloudFormation. My only concern is how I handled `PortMappings` in the `ECSTaskDefinition`; I'm not clear on when/how host ports are selected (random?), and while you'd normally have a parameter such as `MainPort443Host` generated and referenced, I'm not sure how dependent that is on actually having an ELB (which you wouldn't have for UDP or `convox.balancer=false`), so right now it generates

`"443:5000/tcp"`

instead of

```json
{
  "Fn::Join": [
    ":",
    [
      {
        "Ref": "MainPort443Host"
      },
      "5000/tcp"
    ]
  ]
}
```

in those scenarios. Could definitely use some guidance on this. As always, any additional feedback would be appreciated.

**Comment by ddollar** [comment]
It needs to generate the Fn::Join with the Ref in the middle. This is used to select the intermediate instance port (that sits between the ELB and the container port)

**Comment by spencerhakim** [comment]
Sorry if I wasn't clear, that's still the behavior for TCP ports with ELBs enabled (see [here](https://github.com/convox/rack/pull/1535/files#diff-b199ac8aa1c94cf65c1d0181753caf2dR268)), the new behavior I described only affects UDP ports, which can't use ELBs, or TCP ports when `convox.balancer=false`. I have a fixture with both behaviors [here](https://github.com/spencerhakim/rack/blob/048b8bc9387d0888d783ddaffb935261da8bcc15/api/models/fixtures/web_tcp_udp.json#L766).

**Comment by spencerhakim** [comment]
Side note: `make fixtures` currently only generates missing .json files, but the [documentation](https://github.com/convox/rack/blob/master/Development.md#infrastructure-changes) seems to suggest that it should regenerate all of them.

**Comment by spencerhakim** [comment]
So, is there anything else that needs to be done with this? I've removed WIP from the PR name as I think I've done everything I should.

**Comment by ddollar** [comment]
Sorry, have been pushing through some changes lately and looks like this branch needs a rebase. Once you do that I'll review and get it in to our next release.

**Comment by ddollar** [comment]
It also appears that hound has noted some style issues that need to be resolved, mostly around clarifying comments. Let me know if you're not sure how to resolve those.

**Comment by spencerhakim** [comment]
Most of the hound issues are from pre-existing code that I had just moved around. I can add comments to that stuff if you'd like; is it an on-going process in general to add comments to existing code?

**Comment by ddollar** [comment]
Yes, in general we're trying to satisfy the Hound going forward. Let me know if this is too burdensome and I can help.

**Comment by spencerhakim** [comment]
Not a problem, I'm glad to hear that that's an ongoing goal for the codebase and am happy to contribute.

**Comment by spencerhakim** [comment]
Not sure why, but the unit tests started failing after the rebase, both locally and on Travis. I didn't touch the affected tests, and the git logs for the relevant files also don't show any recent changes, so I'm pretty confused.

**Comment by ddollar** [comment]
This may be an issue with master, I'm investigating.

**Comment by spencerhakim** [comment]
@ddollar I figured it out, looks like #1536 [missed something](https://github.com/convox/rack/blob/master/api/models/templates/app.tmpl#L9), plus fixtures need to be regenerated. I can fix it here, or wait for you to fix it in master.

**Comment by spencerhakim** [comment]
Rebased again now that those fixes are in, all the GitHub checks should pass now.

**Comment by ddollar** [comment]
You should be able to safely rebase on master now, it should no longer in flux. Sorry for all the hassle!

**Comment by spencerhakim** [comment]
OK, rebased again, should still be all good.

**Comment by spencerhakim** [comment]
Fixed the issue with installing a rack, and rebased.

**Comment by spencerhakim** [comment]
Rebased and fixed the merge conflict. Any reason we can't get this merged soon?

**Comment by MiguelMoll** [comment]
@spencerhakim Thanks for taking care of this. Any chance you could knock out one more rebase because of the release going out today? That way I can starting to get this out next. Thanks!

**Comment by spencerhakim** [comment]
@MiguelMoll Done.

File: .editorconfig [hunk 3]
```diff
@@ -0,0 +1,19 @@
+root = true
+
```
Comment (MiguelMoll) [inline .editorconfig]: Believe this file was committed by mistake.

**Comment by spencerhakim** [comment]
No, but I can remove it if you'd like, though it's probably a good idea to leave it. http://editorconfig.org/

File: .editorconfig (hunk 3 above)
Comment (spencerhakim) [inline .editorconfig]: No, but I can remove it if you'd like, though it's probably a good idea to leave it. http://editorconfig.org/

Comment (MiguelMoll) [inline .editorconfig]: Yes please that'd be appreciated. As I test this out today, if you like I can tackle any issues that run up. That way I don't keep bothering about little stuff like that.

Comment (spencerhakim) [inline .editorconfig]: Removed. And that's fine with me, it's your project after all.

**Comment by till** [comment]
@MiguelMoll is this merged?

**Comment by MiguelMoll** [comment]
Not yet. It's making its way through our testing and CI process. Hopefully very soon.

**Comment by MiguelMoll** [comment]
Merged via #1639. Thanks again @spencerhakim for this PR! 👍

**Comment by mattmanning** [comment]
[Document UDP support and convox.balancer label](https://trello.com/c/T9XWxY6F/1774-document-udp-support-and-convox-balancer-label)

**Comment by spencerhakim** [comment]
@mattmanning I started on that over [here](https://github.com/convox/site/pull/457). Feel free to use it as a starting point and close the PR.

**Comment by advance512** [comment]
Did this PR break envvar based linking when `convox.balancer=false`?

**Comment by spencerhakim** [comment]
@advance512 Linking depends on a load balancer, so if you explicitly disable the creation of a load balancer then, yes, it breaks linking.

**Comment by advance512** [comment]
Would it not make sense to at least pass one of the created container's endpoint details?

**Comment by spencerhakim** [comment]
@advance512 That doesn't really work due to the somewhat ephemeral nature of containers. Not sure if there's even a clean way Convox could support that if they wanted to. Certainly outside the scope of this PR. If avoiding ELBs is a priority for you, you should look into using a service discovery system like etcd or Consul.

**Comment by till** [comment]
@advance512 we are in the same boat. you need to get creative with service discovery, i think. Did you have any thoughts on this yet?

**Comment by advance512** [comment]
@till 
I am pretty sure many others have been in this same boat, and in fact this might be the "killer feature" that prevents the project from being more popular among small startups and enthusiasts. I have 8 microservices in my project (and this will expand, a microservice per external system I interface into), but only one public access point. That is $140 that I would pay for no reason at all. Too bad, because I really like Convox otherwise.

Anyways, this missing feature might sadly be the reason I will stop using Convox and move to some sort of K8s based solution (like Rancher, Mantl or OpenShift - or many others).

For now, after deployment, I've created a small script that runs `convox api get /apps/APP_NAME/processes`, and updates DNS records in Route53 in which I've created an Private Zone (microservice.app_name.internal). 

Not very fun, definitely not production worthy, but it works.

**Comment by ddollar** [comment]
Stay tuned, we have some good stuff in the works for you folks.

Now that we have host-based routing in ALB we can easily do one-ALB-per-Rack and route the internal hostnames correctly. The team is hard at work getting a completely revamped approach centered around this ready. We're hoping to have something for you all to try in the next few weeks.

**Comment by till** [comment]
@advance512 while @ddollar is working on something, we are playing with this method: https://github.com/convox-examples/dd-agent/pull/3

You basically scale the app to the number of instances and then use something like they did to "discover" the app locally.

**Comment by advance512** [comment]
@till  - that sounds great. Will have a look - though I think it is quickly nearing the same complexity as having a Consul cluster which is what I was trying to avoid.

**Comment by till** [comment]
@advance512 not at all using dd-agent, I am just using/testing their ruby hack for a local statsd that we run.

But yeah, I am also trying to avoid running consul or similar tools.

**Comment by spencerhakim** [comment]
@advance512 For the `dd-agent` example, if you set the label `convox.agent=true` (which scales the process to exactly one instance per EC2 host) and expose `8125/udp`, you can do something like `export EC2_HOSTNAME=$(curl -s http://169.254.169.254/latest/meta-data/local-hostname)` and get a poor man's service discovery by just pointing your app at `${EC2_HOSTNAME}:8125`. `dd-agent` is actually what motivated me to work on this PR and the `convox.agent` PR.

For more info on that `curl` endpoint: http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html

**Comment by advance512** [comment]
I tried using Consul and Linkerd. Let me just say it's very, very complex with Convox.

There are no guides or tutorials on the issue. To create consul cluster, you'd need the addresses of the other servers (TCP, not HTTP), and there is simply no easy way to do it that I could found. You could use `Networking mode: Host` in an ECS task definition to make things easier, and in fact that is what the best article I found [suggests](https://blog.unif.io/deploying-consul-with-ecs-part-2-38de98a56611), but that means you're exiting the Convox world into "direct to AWS" world. You'd probably need to defined DNS records to your Consul servers. Or, you could deploy the Consul servers on their own EC2 instances. However, you'd still need Consul clients and Consul-Registrator on each server. Adding Linkerd for a network mesh puts you in the same problem as with Consul, the need for hard-set peers that are accessible not via ELBs. Then there are issues with docker-compose vs local vs remote - e.g. you need `networks` in the docker-compose to make it work, and it indeed works with Docker Compose but not with `convox start`, you would need to create an exterior network. In production, you'd need `links`. `depends_on` also works in some places and not others.. I don't remember exactly.

I spent nearly 2 days trying to get this to work and eventually just gave up. For now I'm working, again, with multiple ELBs and am actively waiting for the new ALB feature.. if it doesn't arrive soon, I'll have to migrate to Kubernetes. I was trying to avoid this due to not wanting to spend a lot of time on DevOps work at this point in the life of the project.. but I also can't be spending hundreds of dollars per month for ELBs, either.

**Pull request closed without merging**